File: src/java/br/com/lett/crawlernode/crawlers/ranking/keywords/saopaulo/SaopauloBigCrawler.java
package br.com.lett.crawlernode.crawlers.ranking.keywords.saopaulo;

import br.com.lett.crawlernode.core.fetcher.models.Request;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.crawlers.extractionutils.ranking.VTEXRankingKeywords;
import br.com.lett.crawlernode.util.CommonMethods;
import br.com.lett.crawlernode.util.CrawlerUtils;
import br.com.lett.crawlernode.util.JSONUtils;
import br.com.lett.crawlernode.util.Logging;
import org.json.JSONArray;
import org.json.JSONObject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.Base64;

public class SaopauloBigCrawler extends VTEXRankingKeywords {
   
   public SaopauloBigCrawler(Session session) {
      super(session);
   }

   @Override
   protected String getHomePage() {
      return session.getOptions().getString("homePage");
   }

   protected Document fetchDocument() {

      StringBuilder searchPage = new StringBuilder();

      searchPage.append(getHomePage())
         .append(this.keywordEncoded)
         .append("?_q")
         .append(this.keywordEncoded)
         .append("&map=ft&page")
         .append(this.currentPage);

      String apiUrl = searchPage.toString().replace("+", "%20");

      return Jsoup.parse(fetchPage(apiUrl));
   }

   @Override
   protected String getLocation() {
      return "";
   }

   @Override
   protected String getVtexSegment() {
      return session.getOptions().getString("vtex_segment");
   }

}

File: src/java/br/com/lett/crawlernode/crawlers/corecontent/saopaulo/SaopauloBigCrawler.kt
package br.com.lett.crawlernode.crawlers.corecontent.saopaulo

import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.crawlers.extractionutils.core.VTEXNewScraper
import br.com.lett.crawlernode.crawlers.extractionutils.core.VTEXOldScraper
import models.RatingsReviews
import org.apache.http.impl.cookie.BasicClientCookie
import org.json.JSONObject
import org.jsoup.nodes.Document
import java.util.regex.Pattern

class SaopauloBigCrawler(session: Session) : VTEXNewScraper(session) {
   override fun handleCookiesBeforeFetch() {
      val cookie = BasicClientCookie("vtex_segment", session.options.optString("vtex_segment"))
      cookie.domain = if (getHomePage().endsWith("/")) getHomePage().substringAfter("https://").replace("/", "") else getHomePage().substringAfter("https://")
      cookies.add(cookie)
   }

   override fun getHomePage(): String {
      return session.options.optString("homePage")
   }

   override fun getMainSellersNames(): List<String> {
      return session.options?.optJSONArray("sellers")?.toList()
         ?.map { obj: Any -> obj.toString() } ?: listOf()
   }

   override fun scrapRating(internalId: String, internalPid: String, doc: Document, jsonSku: JSONObject): RatingsReviews? {
      return null
   }

   override fun isMainRetailer(sellerName: String?): Boolean {
      var seller = sellerName ?: ""
      if (sellerName != null && sellerName.last() == '.') {
         seller = sellerName.substring(0, sellerName.length - 1)
      }
      return super.isMainRetailer(seller)
   }

   override fun scrapDescription(doc: Document?, productJson: JSONObject?): String {
      val builder = StringBuilder();
      builder.append(productJson?.optString("description")).append("</br></br>Informações do produto</br></br>")

      val keys = productJson?.optJSONArray("allSpecifications")?.toList();

      if (keys != null) {
         for (key in keys) {
            val keyName = key.toString()
            val keyValue = productJson.optJSONArray(keyName)?.toList()?.first()?.toString()
            if (keyValue != null) {
               builder.append(keyName).append(": ").append(keyValue).append("</br>")
            }
         }
      }
      return builder.toString()
   }
}

